phoenix-core/src/main/java/org/apache/phoenix/schema/json/PhoenixJson.java

```diff
@@ -0,0 +1,255 @@
+     */
+    public PhoenixJson getPhoenixJsonOrNull(String[] paths) {
+        try {
+            return getPhoenixJsonInternal(paths);
+        } catch (NumberFormatException nfe) {
+            // ignore
+        }
+        return null;
+    }
+
+    /**
+     * Serialize the current {@link PhoenixJson} to String. Its required for
+     * json_extract_path_text(). If we just return node.toString() it will wrap String value in
+     * double quote which is not the expectation, hence avoiding calling toString() on
+     * {@link JsonNode} until PhoenixJson represent a Json Array or container for Json object. If
+     * PhoenixJson just represent a {@link ValueNode} then it should return value returned from
+     * objects toString().
+     */
+    public String serializeToString() {
+        if (this.rootNode == null || this.rootNode.isNull()) {
+            return null;
+        } else if (this.rootNode.isValueNode()) {
+
+            if (this.rootNode.isNumber()) {
+                return this.rootNode.getNumberValue().toString();
+            } else if (this.rootNode.isBoolean()) {
+                return String.valueOf(this.rootNode.getBooleanValue());
+            } else if (this.rootNode.isTextual()) {
+                return this.rootNode.getTextValue();
+            } else {
+                return getJsonAsString();
+            }
+        } else if (this.rootNode.isArray()) {
+            return getJsonAsString();
+        } else if (this.rootNode.isContainerNode()) {
+            return getJsonAsString();
+        }
+
+        return null;
+
+    }
+
+    @Override
+    public String toString() {
+        return getJsonAsString();
+    }
+
+    @Override
+    public int hashCode() {
+        final int prime = 31;
+        int result = 1;
+        result = prime * result + ((this.rootNode == null) ? 0 : this.rootNode.hashCode());
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) return true;
+        if (obj == null) return false;
+        if (getClass() != obj.getClass()) return false;
+        PhoenixJson other = (PhoenixJson) obj;
+        if ((this.jsonAsString != null) && (other.jsonAsString != null)) {
+            if (this.jsonAsString.equals(other.jsonAsString)) return true;
+        }
+        if (this.rootNode == null) {
+            if (other.rootNode != null) return false;
+        } else if (!this.rootNode.equals(other.rootNode)) return false;
+        return true;
+    }
+
+    /**
+     * @return length of the string represented by the current {@link PhoenixJson}.
+     */
+    public int estimateByteSize() {
+        String jsonStr = getJsonAsString();
+        return jsonStr == null ? 1 : jsonStr.length();
+    }
+
+    public byte[] toBytes() {
+        return Bytes.toBytes(getJsonAsString());
+    }
+
+    @Override
+    public int compareTo(PhoenixJson o) {
+        if (o == null) {
+            return 1;
+        } else if (this.equals(o)) {
```

Comment:
Checking both JsonNode.equals and doing a string compare on the string representation is not so good. Does JSON have the concept of < and > comparison? Should we always rely on string equality to mean node equality? When would that break down? If that works, we should perhaps consistently check that for equality and hashcode, etc. If that doesn't work, we may need to introduce the concept of a type that doesn't support comparison, but only = and !=.